p11-kit/export-object.c

```diff
@@ -59,85 +67,333 @@ int
+	return_val_if_fail (defs, false);
+
+	asn = p11_asn1_create (defs, "PKIX1.SubjectPublicKeyInfo");
+	return_val_if_fail (asn, false);
+
+	result = asn1_write_value (asn, "algorithm.algorithm",
+				   P11_OID_PKIX1_EC_STR, 1);
+	if (result != ASN1_SUCCESS) {
+		p11_message (_("unable to write value"));
+		goto cleanup;
+	}
+
+	result = asn1_write_value (asn, "algorithm.parameters",
+				   ec_params->pValue, ec_params->ulValueLen);
+	if (result != ASN1_SUCCESS) {
+		p11_message (_("unable to write value"));
 		goto cleanup;
 	}
 
+	/* Strip 2 octes of DER header for OCTET STRING */
+	return_val_if_fail (ec_point->ulValueLen > 2, false);
+	result = asn1_write_value (asn, "subjectPublicKey",
+				   ((unsigned char *)ec_point->pValue) + 2,
+				   (ec_point->ulValueLen - 2) * 8);
+	if (result != ASN1_SUCCESS) {
+		p11_message (_("unable to write value"));
+		goto cleanup;
+	}
+
+	der = p11_asn1_encode (asn, &n_der);
+	if (!der) {
+		p11_message (_("unable to encode %s"), "SubjectPublicKeyInfo");
+		goto cleanup;
+	}
+
+	if (!p11_pem_write (der, n_der, "PUBLIC KEY", buf)) {
+		p11_message (_("failed to convert DER to PEM"));
+		goto cleanup;
+	}
+
+	ok = true;
+
 cleanup:
-	p11_buffer_uninit (&buf);
-	free (attr.pValue);
+	free (der);
+	p11_asn1_free (asn);
+	p11_dict_free (defs);
+	p11_attrs_free (attrs);
+	return ok;
 }
 
-static void
-export_certificate (P11KitIter *iter)
+#endif /* WITH_ASN1 */
+
+static bool
+export_pubkey (P11KitIter *iter, p11_buffer *buf)
+{
+	CK_ATTRIBUTE template[] = {
+		{ CKA_PUBLIC_KEY_INFO, },
+		{ CKA_KEY_TYPE, },
+	};
+	CK_ATTRIBUTE *attrs = NULL, *attr;
+	bool ok = false;
+
+	attrs = p11_attrs_buildn (NULL, template, 2);
+	return_val_if_fail (attrs, false);
+
+	if (p11_kit_iter_load_attributes (iter, attrs, p11_attrs_count (attrs)) != CKR_OK) {
+		p11_message (_("failed to retrieve attributes"));
+		goto cleanup;
+	}
+
+	attr = p11_attrs_find_valid (attrs, CKA_PUBLIC_KEY_INFO);
+	if (attr) {
+		if (!p11_pem_write (attr->pValue, attr->ulValueLen, "PUBLIC KEY", buf)) {
+			p11_message (_("failed to convert DER to PEM"));
+			goto cleanup;
+		}
+	} else {
+#ifdef WITH_ASN1
+		CK_KEY_TYPE type;
+
+		if (!p11_attrs_find_ulong (attrs, CKA_KEY_TYPE, &type)) {
+			p11_message (_("unable to determine key type"));
+			goto cleanup;
+		}
+
+		switch (type) {
+		case CKK_RSA:
+			if (!export_pubkey_rsa (iter, buf)) {
+				p11_message (_("unable to extract %s public key"), "RSA");
+				goto cleanup;
+			}
+			break;
+		case CKK_EC:
+			if (!export_pubkey_ec (iter, buf)) {
+				p11_message (_("unable to extract %s public key"), "EC");
+				goto cleanup;
+			}
+			break;
+		default:
+			p11_message (_("unsupported key type"));
+			goto cleanup;
+		}
+#else /* WITH_ASN1 */
+		p11_message (_("ASN.1 support is not compiled in"));
+		goto cleanup;
+#endif /* !WITH_ASN1 */
+	}
+
+	ok = true;
+
+cleanup:
+	p11_attrs_free (attrs);
+	return ok;
+}
+
+static bool
+export_certificate (P11KitIter *iter, p11_buffer *buf)
```

Comment:
each function parameter should be on a separate line judging by the code format of other commands.